app/src/main/java/us/shandian/giga/service/DownloadManagerService.java

```diff
@@ -380,16 +381,17 @@ public static void startMission(Context context, String[] urls, StoredFileHelper
 
     private void startMission(Intent intent) {
         String[] urls = intent.getStringArrayExtra(EXTRA_URLS);
-        Uri path = intent.getParcelableExtra(EXTRA_PATH);
-        Uri parentPath = intent.getParcelableExtra(EXTRA_PARENT_PATH);
+        Uri path = IntentCompat.getParcelableExtra(intent, EXTRA_PATH, Uri.class);
+        Uri parentPath = IntentCompat.getParcelableExtra(intent, EXTRA_PARENT_PATH, Uri.class);
         int threads = intent.getIntExtra(EXTRA_THREADS, 1);
         char kind = intent.getCharExtra(EXTRA_KIND, '?');
         String psName = intent.getStringExtra(EXTRA_POSTPROCESSING_NAME);
         String[] psArgs = intent.getStringArrayExtra(EXTRA_POSTPROCESSING_ARGS);
         String source = intent.getStringExtra(EXTRA_SOURCE);
         long nearLength = intent.getLongExtra(EXTRA_NEAR_LENGTH, 0);
         String tag = intent.getStringExtra(EXTRA_STORAGE_TAG);
-        Parcelable[] parcelRecovery = intent.getParcelableArrayExtra(EXTRA_RECOVERY_INFO);
+        Parcelable[] parcelRecovery = IntentCompat.getParcelableArrayExtra(
+                intent, EXTRA_RECOVERY_INFO, Parcelable.class);
```

Comment:
```suggestion
        List<MissionRecoveryInfo> recovery = IntentCompat.getParcelableArrayListExtra(
                intent, EXTRA_RECOVERY_INFO, MissionRecoveryInfo.class);
```

The above suggestion would require some additional changes to avoid issues, e.g. changing the array in `AboutActivity` to a list.

Reply:
Could you go into more detail?

Reply:
Could you go into more detail?

Reply:
Yeah, the Parcelable array method requires casting, but the list method does not.

Reply:
Could you pull the branch and make the changes, please? I don't really have the time at the moment. I allow edits by maintainers.